### PR TITLE
Add logging to predicates

### PR DESCRIPTION
**Pull request by orf**
One thing that would be really helpful as you're developing an app that uses django-rules is to have some form of logging output as to the results of nested predicates.

For example in our app we have some fairly complex predicates that are nested, and if the whole chain returns `False` it would be handy to know exactly what triggered that without having to dive into all of the interactions in our `rules.py`. 

This MR adds initial support for this, it will need some more work but please let me know if you think I should spend more time fleshing this out.

The logging output looks like this when running the test suite:
```
Testing ((requires_two_args & requires_two_args) & passthrough)
requires_two_args = skipped
requires_two_args = skipped
(requires_two_args & requires_two_args) = skipped
passthrough = True
((requires_two_args & requires_two_args) & passthrough) = True
```

And

```
Testing (passthrough | ~requires_two_args)
passthrough = True
(passthrough | ~requires_two_args) = True
```

Edit: I didn't handle skipped predicates, I added a commit to do this.

### CODE REVIEW

**Comment by dfunckt** [comment]
I like this 👍 Please do flesh it out as you like and I'll happily review and merge.

Some initial feedback:

1. I'd prefer the log level to be less than `info` since it's not informative output, but debug output. I can't really remember how logging config works exactly, but it would be great if we saved users the need to configure a logger for rules unless they really need to debug rules.
2. Eventually add some docs for the feature, maybe show an example config -- a Django-related example only should be enough I think.

Thanks for the PR, looks great!

**Comment by danielnaab** [comment]
This seems like a good idea - but how about add logging that creates an inspectable object with the results of the predicate evaluation, rather than just logging strings? Then a separate function could push that object to a logger, or code could be written to inspect the object in other manners.

**Comment by orf** [comment]
I think currently thats slightly over-engineering it to be honest @danielnaab. I think that would be good to have, but it would need a fair bit of refactoring. Right now as far as I can tell each individual predicate has no idea about it's parent which would be needed if we go down that route.

Also logging is essentially does nothing (not even the string formatting) if it's not needed, whereas something like that would always have a performance impact I think?

Feel free to take a stab at it though, I'll finish up this MR and then we can see if we can improve it :+1:

**Comment by orf** [comment]
@dfunckt I've made those changes and added information to the readme. Let me know if you think it's OK, and if you merge it make sure to squash the commits :gun:

File: README.rst [hunk 1]
```diff
@@ -623,6 +624,37 @@ leaving the predicate result up to that point unchanged.
 older versions by calling the predicate's ``skip()`` method, but this has been
 deprecated and support will be completely removed in a future version.
 
+
+Logging predicate evaluation
+----------------------------
+
+Complex or heavily nested rules can be painful to debug. This package provides a `django-rules` logger that can be configured in your django settings to output the results of individual evaluations to the console. Messages are sent at the `DEBUG` level.
```
Comment (dfunckt) [inline README.rst]: Can you please break lines at around 80 columns?

Comment (dfunckt) [inline README.rst]: I would rephrase to something along these lines:

> `rules` can optionally be configured to log debug information as rules are evaluated to help with debugging your predicates. Messages are sent at the `DEBUG` level. The following [`dictConfig`](https://docs.python.org/3.6/library/logging.config.html#logging-config-dictschema) configures the logger to log to the console: ...

File: rules/predicates.py [hunk 2]
```diff
@@ -1,10 +1,14 @@
 import inspect
 import operator
 import threading
+import logging
 from functools import partial, update_wrapper
 from warnings import warn
 
 
+logger = logging.getLogger('django-rules')
```
Comment (dfunckt) [inline rules/predicates.py]: Let's rename the logger to 'rules'.

File: rules/predicates.py [hunk 3]
```diff
@@ -211,9 +216,12 @@ def _apply(self, *args):
             callargs = (self,) + callargs
         try:
             result = self.fn(*callargs)
-            return None if result is None else bool(result)
+            returner = None if result is None else bool(result)
```
Comment (dfunckt) [inline rules/predicates.py]: It's safe to overwrite `result`. "returner" feels somewhat awkward as an identifier here.

File: rules/predicates.py [hunk 4]
```diff
@@ -211,9 +216,12 @@ def _apply(self, *args):
...
         except SkipPredicate:
-            return None
+            returner = None
+        
+        logger.debug('%s = %s', self, 'skipped' if returner is None else returner)
```
Comment (dfunckt) [inline rules/predicates.py]: I think it would be nice to indent by two spaces this one, so that output is visually parsed as a group:

```
Testing ((requires_two_args & requires_two_args) & passthrough)
  requires_two_args = skipped
  requires_two_args = skipped
  (requires_two_args & requires_two_args) = skipped
  passthrough = True
  ((requires_two_args & requires_two_args) & passthrough) = True
```

I would also change the equals sign to `=>` or anything you feel that can convey direction, i.e. "this resulted in that".

File: tests/testapp/settings.py [hunk 5]
```diff
@@ -58,3 +58,21 @@
         },
     },
 ]
+
+LOGGING = {
```
Comment (dfunckt) [inline tests/testapp/settings.py]: I don't see great value having the logger enabled in tests where everything is under control (i.e. we're not debugging) so I'd rather we omitted it because it messes with test output.

**Comment by orf** [comment]
I've implemented those changes @dfunckt, let me know if it all checks out 🎱

**Comment by dfunckt** [comment]
Manually merged in 27960712 with minor modifications -- big thanks!
